rewriter/src/main/java/com/facebook/coresql/rewriter/OrderByPatternMatcher.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.coresql.rewriter;
+
+import com.facebook.coresql.parser.AstNode;
+import com.facebook.coresql.parser.QuerySpecification;
+import com.facebook.coresql.parser.SimpleNode;
+import com.facebook.coresql.parser.SqlParserDefaultVisitor;
+import com.facebook.coresql.parser.Subquery;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTINSERT;
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTLIMITCLAUSE;
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTORDERBYCLAUSE;
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTTABLEDEFINITION;
+import static java.util.Objects.requireNonNull;
+
+public class OrderByPatternMatcher
+        extends SqlParserDefaultVisitor
+        implements PatternMatcher
+{
+    private final Set<AstNode> matchedNodes;
+    private int depth;
+    private static final int MINIMUM_SUBQUERY_DEPTH = 2; // Past this depth, all queries we encounter are subqueries
+
+    public OrderByPatternMatcher()
+    {
+        matchedNodes = new HashSet<>();
+    }
+
+    @Override
+    public Set<AstNode> findMatchedNodes(AstNode node)
+    {
+        requireNonNull(node, "AST passed to pattern matcher was null");
+        node.jjtAccept(this, null);
+        return matchedNodes;
+    }
+
+    /**
+     * Checks for a specific pattern in the AST this pattern matcher is traversing:
+     * 1. Query has an ORDER BY
+     * 2. Query does not have a LIMIT
+     * 3. The result of the query is used to CREATE a table or INSERT into a table
+     *
+     * @param node The node we're currently visiting
+     * @return True if the pattern is matched at the current node, else false
+     */
+    private boolean patternIsMatched(AstNode node)
```

Comment:
This is a private method so it makes more sense to give it a meaningful name rather than generic. Maybe use `isOrderByWithoutLimitSubQuery` or something along that line.

Reply:
Agreed.

Done